File: backend/api/routers/devices.py
"""
IMS 2.0 - Login-device enrolment + approval (owner rulings 2026-09-02)
=======================================================================
Staff sign-in is restricted to SUPERADMIN-approved devices (see
api/services/device_gate.py -- DARK until DEVICE_GATE_MODE is armed).

This router is INCLUDED BY auth.py under its own prefix, so everything here
lives at /api/v1/auth/devices/* -- the device gate is part of authentication,
and mounting it from auth.py keeps main.py untouched.

Route map (POLICY rows in api/services/rbac_policy.py):
  POST /auth/devices/enroll/options     PUBLIC*  challenge for create()
  POST /auth/devices/enroll             PUBLIC*  submit a PENDING request
  POST /auth/devices/assertion-options  PUBLIC   challenge for get()
  GET  /auth/devices                    SUPERADMIN  list (approval screen)
  POST /auth/devices/{id}/approve       SUPERADMIN
  POST /auth/devices/{id}/revoke        SUPERADMIN  (also rejects a PENDING row)

* "PUBLIC" = no session token, but BOTH enrolment routes demand a valid
  ENROLMENT TICKET -- a 10-minute purpose-scoped JWT that /auth/login mints
  only after verifying a correct password on a gated account. Nobody can
  enrol (or spam PENDING rows) without first proving a real password, and
  the ticket is refused as a session token by get_current_user.

Approval is SUPERADMIN-ONLY by owner ruling: an ADMIN is exempt from the
gate but cannot enrol anybody's device. Do not widen this -- the answer to
"we need a second approver" is a second SUPERADMIN account.
"""

from __future__ import annotations

import secrets
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from api.services import device_gate

# Imported at the TOP deliberately. auth.py includes this router at its
# BOTTOM (after these names are defined), so the normal import order --
# main -> auth -> devices -> (auth already in sys.modules) -- is safe. If
# anything ever imports THIS module before auth, the include in auth.py
# raises a loud ImportError instead of silently registering half a router.
from .auth import (  # noqa: E402
    _audit_auth_event,
    _client_ip,
    get_current_user as _get_current_user,
    require_roles,
)

router = APIRouter()

_superadmin = require_roles("SUPERADMIN")


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


class EnrollOptionsRequest(BaseModel):
    enroll_ticket: str


class EnrollRequest(BaseModel):
    enroll_ticket: str
    challenge_id: str
    credential_id: str = Field(..., min_length=1)
    client_data_json: str
    # SPKI DER from AuthenticatorAttestationResponse.getPublicKey(), base64.
    public_key_spki: str = Field(..., min_length=1)
    public_key_alg: Optional[int] = None  # COSE alg (-7 ES256 / -257 RS256)
    # Raw getAuthenticatorData() bytes (base64) -- only used to read the
    # backup-eligible flag; optional for older browsers.
    authenticator_data: Optional[str] = None
    device_name: str = Field(..., min_length=1, max_length=80)
    platform: Optional[str] = None


def _require_ticket(ticket: str) -> dict:
    claims = device_gate.read_enroll_ticket(ticket)
    if claims is None:
        raise HTTPException(
            status_code=401,
            detail="Enrolment ticket invalid or expired. Sign in again to restart device registration.",
        )
    return claims


# ---------------------------------------------------------------------------
# Pre-auth: enrolment + login-assertion challenges
# ---------------------------------------------------------------------------


@router.post("/enroll/options")
async def enroll_options(request: EnrollOptionsRequest):
    """Server half of navigator.credentials.create(): challenge + RP info.
    Ticket-gated (a correct password was just presented)."""
    claims = _require_ticket(request.enroll_ticket)
    options = device_gate.new_challenge()
    options["user_handle"] = secrets.token_urlsafe(16)
    options["username"] = claims.get("username") or ""
    return options


@router.post("/enroll")
async def enroll_device(request: EnrollRequest, req: Request = None):
    """Record a PENDING device. Enrolment is an ACT, never a side effect of a
    login -- a human typed a device name and a SUPERADMIN must still approve."""
    claims = _require_ticket(request.enroll_ticket)
    doc, reason = device_gate.create_enrolment(
        ticket_claims=claims,
        challenge_id=request.challenge_id,
        credential_id=request.credential_id,
        client_data_json=request.client_data_json,
        public_key_spki=request.public_key_spki,
        public_key_alg=request.public_key_alg,
        authenticator_data=request.authenticator_data,
        device_name=request.device_name,
        platform=request.platform,
    )
    if doc is None:
        raise HTTPException(status_code=400, detail=f"Enrolment failed: {reason}")
    _audit_auth_event(
        action="device_enroll_requested",
        user_id=claims.get("user_id"),
        username=claims.get("username"),
        ip_address=_client_ip(req),
        severity="INFO",
        detail=f"{doc['device_id']} '{doc['device_name']}'",
    )
    return {
        "device_id": doc["device_id"],
        "status": doc["status"],
        "message": "Device registered. Awaiting SUPERADMIN approval.",
    }


@router.post("/enroll-ticket")
async def enroll_ticket_for_current_session(current_user: dict = Depends(_get_current_user)):
    """AUTHENTICATED pre-arming path. While the gate is off (dark) or in log
    mode, staff still sign in normally -- this mints an enrolment ticket for
    the CURRENT session so store devices can be registered and approved
    BEFORE DEVICE_GATE_MODE=enforce is armed. Without it, the only way to
    get a ticket is an enforce-mode rejection, i.e. arming the gate would
    block every till at once on day one. The ticket is identical to the one
    a rejected login mints (same purpose scoping, same 10-minute life)."""
    return {
        "enroll_ticket": device_gate.mint_enroll_ticket(
            current_user.get("user_id") or "",
            current_user.get("username") or "",
        )
    }


@router.post("/assertion-options")
async def assertion_options():
    """Fresh single-use challenge for navigator.credentials.get() before a
    gated login. Deliberately anonymous: it reveals nothing (no credential
    ids are served -- the browser keeps those) and a challenge is just a
    TTL'd random value in the shared cache."""
    return device_gate.new_challenge()


# ---------------------------------------------------------------------------
# SUPERADMIN management (the phone-browser approval screen)
# ---------------------------------------------------------------------------

@router.get("")
async def list_devices(current_user: dict = Depends(_superadmin)):
    """All device rows (pending first) + the live gate mode, so the approval
    screen can say plainly whether the gate is OFF (dark), LOG, or ENFORCE."""
    return {"mode": device_gate.gate_mode(), "devices": device_gate.list_devices()}


@router.post("/{device_id}/approve")
async def approve_device(
    device_id: str,
    req: Request = None,
    current_user: dict = Depends(_superadmin),
):
    doc = device_gate.set_device_status(device_id, "APPROVED", current_user)
    if doc is None:
        raise HTTPException(status_code=404, detail="Device not found")
    _audit_auth_event(
        action="device_approved",
        user_id=current_user.get("user_id"),
        username=current_user.get("username"),
        ip_address=_client_ip(req),
        severity="INFO",
        detail=f"{device_id} '{doc.get('device_name')}'",
    )
    return doc


@router.post("/{device_id}/revoke")
async def revoke_device(
    device_id: str,
    req: Request = None,
    current_user: dict = Depends(_superadmin),
):
    doc = device_gate.set_device_status(device_id, "REVOKED", current_user)
    if doc is None:
        raise HTTPException(status_code=404, detail="Device not found")
    _audit_auth_event(
        action="device_revoked",
        user_id=current_user.get("user_id"),
        username=current_user.get("username"),
        ip_address=_client_ip(req),
        severity="WARNING",
        detail=f"{device_id} '{doc.get('device_name')}'",
    )
    return doc

File: frontend/src/services/api/devices.ts
// ============================================================================
// IMS 2.0 - Approved-device login gate API (owner rulings 2026-09-02)
// ============================================================================
// Client for backend/api/routers/devices.py (/auth/devices/*). The gate is
// DARK until the backend env DEVICE_GATE_MODE is armed; every call here is
// safe to make while it is off.
//
// Two halves:
//   * devicesApi        - SUPERADMIN management (the phone approval screen)
//                         + the pre-auth enrolment/assertion endpoints.
//   * WebAuthn helpers  - the browser passkey ceremonies the LOGIN PAGE will
//                         wire once the gate is armed (enrolDevicePasskey /
//                         getDeviceAssertion). They live here so the login
//                         wiring is a single call, and so the base64 plumbing
//                         has exactly one implementation.
//
// Import directly from this module (not the barrel) per convention.

import api from './client';

// ----------------------------------------------------------------------------
// Types
// ----------------------------------------------------------------------------

export type DeviceStatus = 'PENDING' | 'APPROVED' | 'REVOKED';

export interface LoginDevice {
  device_id: string;
  credential_id: string;
  status: DeviceStatus | string;
  device_name: string;
  platform?: string | null;
  /** True when the passkey is sync-eligible (iCloud / Google Password
   *  Manager) and may follow the enrolling person to their other devices —
   *  shown as a warning chip on the approval screen. Null = unknown. */
  backup_eligible?: boolean | null;
  requested_by?: { user_id?: string; username?: string } | null;
  requested_at?: string | null;
  approved_by?: { user_id?: string; username?: string } | null;
  approved_at?: string | null;
  revoked_by?: { user_id?: string; username?: string } | null;
  revoked_at?: string | null;
}

export interface DeviceListResponse {
  /** Live gate mode: 'off' (dark), 'log' (dry run), 'enforce'. */
  mode: 'off' | 'log' | 'enforce' | string;
  devices: LoginDevice[];
}

export interface ChallengeOptions {
  challenge_id: string;
  challenge: string; // base64url
  rp_id: string;
  timeout_ms?: number;
  user_handle?: string; // enrolment only
  username?: string; // enrolment only
}

/** Shape of LoginRequest.device_assertion on POST /auth/login. */
export interface DeviceAssertionPayload {
  challenge_id: string;
  credential_id: string;
  client_data_json: string;
  authenticator_data: string;
  signature: string;
}

// ----------------------------------------------------------------------------
// API
// ----------------------------------------------------------------------------

export const devicesApi = {
  /** SUPERADMIN: all device rows (pending first) + live gate mode. */
  list: async (): Promise<DeviceListResponse> => {
    const { data } = await api.get<DeviceListResponse>('/auth/devices');
    return data;
  },

  /** SUPERADMIN: approve a pending (or re-approve a revoked) device. */
  approve: async (deviceId: string): Promise<LoginDevice> => {
    const { data } = await api.post<LoginDevice>(
      `/auth/devices/${deviceId}/approve`
    );
    return data;
  },

  /** SUPERADMIN: revoke an approved device / reject a pending request. */
  revoke: async (deviceId: string): Promise<LoginDevice> => {
    const { data } = await api.post<LoginDevice>(
      `/auth/devices/${deviceId}/revoke`
    );
    return data;
  },

  /** AUTHENTICATED pre-arming path: while the gate is off (dark) or in
   *  dry-run, staff still sign in normally — this mints a ticket so the
   *  CURRENT device can be registered and approved BEFORE enforce mode is
   *  armed. Without it, arming would block every till at once. */
  enrollTicket: async (): Promise<{ enroll_ticket: string }> => {
    const { data } = await api.post<{ enroll_ticket: string }>(
      '/auth/devices/enroll-ticket'
    );
    return data;
  },

  /** Pre-auth: challenge for the create() ceremony (needs the enrolment
   *  ticket a device-rejected login returns). */
  enrollOptions: async (enrollTicket: string): Promise<ChallengeOptions> => {
    const { data } = await api.post<ChallengeOptions>(
      '/auth/devices/enroll/options',
      { enroll_ticket: enrollTicket }
    );
    return data;
  },

  /** Pre-auth: submit the created credential as a PENDING device. */
  enroll: async (payload: {
    enroll_ticket: string;
    challenge_id: string;
    credential_id: string;
    client_data_json: string;
    public_key_spki: string;
    public_key_alg?: number | null;
    authenticator_data?: string | null;
    device_name: string;
    platform?: string;
  }): Promise<{ device_id: string; status: string; message: string }> => {
    const { data } = await api.post('/auth/devices/enroll', payload);
    return data;
  },

  /** Pre-auth: fresh single-use challenge for the get() ceremony. */
  assertionOptions: async (): Promise<ChallengeOptions> => {
    const { data } = await api.post<ChallengeOptions>(
      '/auth/devices/assertion-options'
    );
    return data;
  },
};

// ----------------------------------------------------------------------------
// WebAuthn browser ceremonies (for the login-page wiring, once armed)
// ----------------------------------------------------------------------------

/** localStorage key holding this browser's enrolled credential id (NOT a
 *  secret — the private key never leaves the platform authenticator). */
export const DEVICE_CREDENTIAL_KEY = 'ims_device_credential_id';

const b64ToBytes = (value: string): Uint8Array => {
  const pad = value.replace(/-/g, '+').replace(/_/g, '/');
  const raw = atob(pad + '='.repeat((4 - (pad.length % 4)) % 4));
  return Uint8Array.from(raw, (c) => c.charCodeAt(0));
};

const bytesToB64 = (buf: ArrayBuffer): string => {
  let out = '';
  const bytes = new Uint8Array(buf);
  for (let i = 0; i < bytes.length; i++) out += String.fromCharCode(bytes[i]);
  return btoa(out);
};

const bytesToB64Url = (buf: ArrayBuffer): string =>
  bytesToB64(buf).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');

/** Create a platform passkey for THIS device and submit it for approval.
 *  Returns the pending device id. Throws if the browser lacks WebAuthn or
 *  the user cancels the OS prompt. */
export async function enrolDevicePasskey(
  enrollTicket: string,
  deviceName: string
): Promise<{ device_id: string; status: string; message: string }> {
  const opts = await devicesApi.enrollOptions(enrollTicket);
  const credential = (await navigator.credentials.create({
    publicKey: {
      challenge: b64ToBytes(opts.challenge) as BufferSource,
      rp: { id: opts.rp_id || undefined, name: 'IMS 2.0' },
      user: {
        id: b64ToBytes(opts.user_handle || opts.challenge_id) as BufferSource,
        name: deviceName,
        displayName: deviceName,
      },
      pubKeyCredParams: [
        { type: 'public-key', alg: -7 }, // ES256
        { type: 'public-key', alg: -257 }, // RS256 (Windows Hello)
      ],
      authenticatorSelection: {
        authenticatorAttachment: 'platform',
        residentKey: 'preferred',
        userVerification: 'required',
      },
      attestation: 'none',
      timeout: opts.timeout_ms || 60000,
    },
  })) as PublicKeyCredential | null;
  if (!credential) throw new Error('Device registration was cancelled');
  const response = credential.response as AuthenticatorAttestationResponse;
  const spki = response.getPublicKey?.();
  if (!spki) throw new Error('This browser cannot register a device passkey');
  const result = await devicesApi.enroll({
    enroll_ticket: enrollTicket,
    challenge_id: opts.challenge_id,
    credential_id: credential.id,
    client_data_json: bytesToB64(response.clientDataJSON),
    public_key_spki: bytesToB64(spki),
    public_key_alg: response.getPublicKeyAlgorithm?.() ?? null,
    authenticator_data: response.getAuthenticatorData
      ? bytesToB64(response.getAuthenticatorData())
      : null,
    device_name: deviceName,
    platform: navigator.userAgent,
  });
  try {
    localStorage.setItem(DEVICE_CREDENTIAL_KEY, credential.id);
  } catch {
    /* private mode - re-enrolment will be offered */
  }
  return result;
}

/** Sign a fresh server challenge with this device's passkey; the result goes
 *  on LoginRequest.device_assertion. Returns null when this browser has no
 *  enrolled credential recorded (caller then shows the enrol flow). */
export async function getDeviceAssertion(): Promise<DeviceAssertionPayload | null> {
  let credentialId: string | null = null;
  try {
    credentialId = localStorage.getItem(DEVICE_CREDENTIAL_KEY);
  } catch {
    credentialId = null;
  }
  if (!credentialId) return null;
  const opts = await devicesApi.assertionOptions();
  const credential = (await navigator.credentials.get({
    publicKey: {
      challenge: b64ToBytes(opts.challenge) as BufferSource,
      rpId: opts.rp_id || undefined,
      allowCredentials: [
        { type: 'public-key', id: b64ToBytes(credentialId) as BufferSource },
      ],
      userVerification: 'preferred',
      timeout: opts.timeout_ms || 60000,
    },
  })) as PublicKeyCredential | null;
  if (!credential) return null;
  const response = credential.response as AuthenticatorAssertionResponse;
  return {
    challenge_id: opts.challenge_id,
    credential_id: bytesToB64Url(credential.rawId),
    client_data_json: bytesToB64(response.clientDataJSON),
    authenticator_data: bytesToB64(response.authenticatorData),
    signature: bytesToB64(response.signature),
  };
}
